File: cli/curriculum/build-index.mjs
#!/usr/bin/env node
// Usage: node cli/curriculum/build-index.mjs <nfo-root-dir> <show-id> <out.json>
// Reads Season */*.nfo under <nfo-root-dir>, writes the merged index JSON.
import { readFileSync, writeFileSync, readdirSync, statSync } from 'fs';
import { join } from 'path';
import { parseEpisodeNfo, parseSeasonNfo, buildIndex } from './nfoIndex.mjs';

// Authored season → category map (see spec). Adjust here if categories change.
const SEASON_META = {
  0: { title: 'Reference', category: 'reference', pinned: true },
  1: { category: 'lesson', sequential: true }, 2: { category: 'lesson', sequential: true },
  3: { category: 'lesson', sequential: true }, 4: { category: 'lesson', sequential: true },
  5: { category: 'lesson', sequential: true }, 6: { category: 'lesson', sequential: true },
  7: { category: 'lesson', sequential: true }, 8: { category: 'lesson', sequential: true },
  9: { category: 'lesson', sequential: true },
  10: { category: 'repertoire', kind: 'tutorial', facets: ['difficulty', 'instructor', 'style'] },
  11: { category: 'repertoire', kind: 'challenge', facets: ['difficulty', 'instructor', 'style'] },
  12: { category: 'repertoire', kind: 'accompaniment', facets: ['difficulty', 'instructor', 'style'] },
};

const [root, showId, out] = process.argv.slice(2);
if (!root || !showId || !out) { console.error('Usage: build-index.mjs <nfo-root> <show-id> <out.json>'); process.exit(1); }

const episodes = [];
const seasonTitles = {};
for (const dir of readdirSync(root)) {
  const p = join(root, dir);
  if (!statSync(p).isDirectory() || !/^Season /.test(dir)) continue;
  for (const f of readdirSync(p)) {
    if (!f.endsWith('.nfo')) continue;
    const xml = readFileSync(join(p, f), 'utf8');
    if (f === 'season.nfo') { const s = parseSeasonNfo(xml); if (s.season != null) seasonTitles[s.season] = s.title; continue; }
    const ep = parseEpisodeNfo(xml); if (ep) episodes.push(ep);
  }
}
// Merge scanned season titles into the authored meta.
const seasonMeta = {};
for (const [sn, meta] of Object.entries(SEASON_META)) seasonMeta[sn] = { title: seasonTitles[sn] ?? meta.title, ...meta };

const idx = buildIndex({ show: Number(showId), seasonMeta, episodes });
writeFileSync(out, JSON.stringify(idx, null, 1));
console.log(`wrote ${out}: ${Object.keys(idx.seasons).length} seasons, ${Object.keys(idx.episodes).length} episodes`);

File: cli/curriculum/build-index.py
import os, re, json, glob
ROOT = "/media/kckern/Media/Lectures/Piano With Jonny"
GENERIC = {"Music", "Educational"}

# authored season category map (from the design)
SEASON_CAT = {
 0:  dict(category="reference", pinned=True),
 1:  dict(category="lesson", sequential=True),
 2:  dict(category="lesson", sequential=True),
 3:  dict(category="lesson", sequential=True),
 4:  dict(category="lesson", sequential=True),
 5:  dict(category="lesson", sequential=True),
 6:  dict(category="lesson", sequential=True),
 7:  dict(category="lesson", sequential=True),
 8:  dict(category="lesson", sequential=True),
 9:  dict(category="lesson", sequential=True),
 10: dict(category="repertoire", kind="tutorial",      facets=["difficulty","instructor","style"]),
 11: dict(category="repertoire", kind="challenge",     facets=["difficulty","instructor","style"]),
 12: dict(category="repertoire", kind="accompaniment", facets=["difficulty","instructor","style"]),
}

def tag(txt, key):
    m = re.findall(r"<tag>%s:\s*([^<]+)</tag>" % re.escape(key), txt)
    return [x.strip() for x in m]

def one(txt, el):
    m = re.search(r"<%s>(.*?)</%s>" % (el, el), txt, re.S)
    return (m.group(1).strip() if m else None)

def unesc(s):
    if s is None: return None
    return s.replace("&amp;","&").replace("&lt;","<").replace("&gt;",">").replace("&#39;","'").replace("&quot;",'"')

episodes = {}
season_titles = {}
counts = {}
for nfo in glob.glob(os.path.join(ROOT, "Season *", "*.nfo")):
    base = os.path.basename(nfo)
    if base == "season.nfo":
        txt = open(nfo, encoding="utf-8").read()
        sn = one(txt, "seasonnumber"); ti = one(txt, "title")
        if sn is not None: season_titles[int(sn)] = unesc(ti)
        continue
    txt = open(nfo, encoding="utf-8").read()
    s = one(txt, "season"); e = one(txt, "episode")
    if s is None or e is None: continue
    s, e = int(s), int(e)
    genres = re.findall(r"<genre>([^<]+)</genre>", txt)
    style = next((g for g in genres if g not in GENERIC), None)
    ep = dict(
        title=unesc(one(txt, "title")),
        plot=unesc(one(txt, "plot")),
        course=unesc((tag(txt,"Course") or [None])[0]),
        style=unesc(style),
        skill=(tag(txt,"Skill Level") or [None])[0],
        focus=tag(txt,"Focus") or [],
        type=(tag(txt,"Type") or [None])[0],
        instructor=unesc(one(txt,"credits")),
    )
    episodes["%d:%d" % (s,e)] = {k:v for k,v in ep.items() if v not in (None,[],"")}
    counts[s] = counts.get(s,0)+1

seasons = {}
for sn, meta in SEASON_CAT.items():
    seasons[sn] = dict(title=season_titles.get(sn), episodes=counts.get(sn,0), **meta)

out = dict(show=676490, seasons=seasons, episodes=episodes)
open("%s/676490.index.json" % os.environ["SCRATCH"], "w").write(json.dumps(out, ensure_ascii=False, indent=1))
print("seasons:", len(seasons), "episodes:", len(episodes))
print("sample 10:1 =", json.dumps(episodes.get("10:1"), ensure_ascii=False))
print("sample 1:1  =", json.dumps(episodes.get("1:1"), ensure_ascii=False))
print("distinct styles:", sorted({v.get("style") for v in episodes.values() if v.get("style")}))
